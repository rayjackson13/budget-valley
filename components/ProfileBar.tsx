import React, { useEffect, useRef, useState } from 'react';
import { Animated, Easing, Image, LayoutChangeEvent, StyleSheet, View, ViewStyle } from 'react-native';
import { Text, Title, useTheme } from 'react-native-paper';
import ProfileImage from 'assets/images/profile.jpeg';

type ProfileBarProps = {
  style?: ViewStyle
}

export default function ProfileBar({ style }: ProfileBarProps) {
  const [viewWidth, setViewWidth] = useState(0);
  const fadeInAnim = useRef(new Animated.Value(0)).current;
  const theme = useTheme();
  const styles = useStyles(theme);
  const viewPadding = 5;
  const avatarSize = 50;
  const translateX = fadeInAnim.interpolate({
    inputRange: [0, 1],
    outputRange: [viewWidth - avatarSize - viewPadding * 2, 0],
  });
  const rotate = fadeInAnim.interpolate({
    inputRange: [0, 1],
    outputRange: ['0deg', '360deg'],
  });

  const onLayout = (event: LayoutChangeEvent) => {
    const { layout } = event.nativeEvent;
    setViewWidth(layout.width);
  };

  const runFadeInAnimation = () => {
    Animated.timing(fadeInAnim, {
      useNativeDriver: true,
      duration: 600,
      toValue: 1,
      easing: Easing.bezier(0.16, 1, 0.3, 1),
    }).start();
  };

  useEffect(() => {
    runFadeInAnimation();
  }, []);

  return (
    <Animated.View
      style={[styles.wrap, style, { padding: viewPadding, opacity: fadeInAnim }]}
      onLayout={onLayout}
    >
      <View style={styles.row}>
        <Animated.View style={[styles.avatar, { transform: [{ translateX }, { rotate }], opacity: fadeInAnim }]}>
          <Image source={ProfileImage} style={styles.image} resizeMode="cover" />
        </Animated.View>
        <Animated.View style={[styles.content, { opacity: fadeInAnim }]}>
          <Title style={styles.title}>
            {'Welcome back, '}
            <Text style={styles.accent}>Kostya</Text>
            !
          </Title>
        </Animated.View>
      </View>
    </Animated.View>
  );
}

const useStyles = ({ colors }: ReactNativePaper.Theme) => StyleSheet.create({
  wrap: {
    width: '100%',
    padding: 5,
    backgroundColor: colors.darkGrey,
    borderRadius: 64,
  },
  row: {
    flexDirection: 'row',
  },
  avatar: {
    width: 50,
    height: 50,
  },
  image: {
    width: '100%',
    height: '100%',
    borderRadius: 50,
  },
  content: {
    flex: 1,
    paddingHorizontal: 15,
    paddingRight: 20,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    textAlign: 'center',
  },
  accent: {
    color: colors.accent,
    fontWeight: 'bold',
  },
});
