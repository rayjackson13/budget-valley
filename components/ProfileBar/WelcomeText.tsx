import React from 'react';
import { StyleSheet } from 'react-native';
import { Text, Title } from 'react-native-paper';

import type { UserData } from '.';

type WelcomeTextPropsType = {
  user: UserData;
  theme: ReactNativePaper.Theme;
}

const GREETINGS = 'Welcome back';

export default function WelcomeText({ user, theme }: WelcomeTextPropsType) {
  const styles = useStyles(theme);
  if (!user || !user.name) {
    return (
      <Title style={styles.title}>
        {GREETINGS}
        !
      </Title>
    );
  }

  return (
    <Title style={styles.title}>
      {`${GREETINGS}, `}
      <Text style={styles.accent} testID="username">{user.name}</Text>
      !
    </Title>
  );
}

const useStyles = ({ colors }: ReactNativePaper.Theme) => StyleSheet.create({
  title: {
    textAlign: 'center',
  },
  accent: {
    color: colors.accent,
    fontWeight: 'bold',
  },
});
