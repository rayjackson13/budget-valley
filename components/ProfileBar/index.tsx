import React, { useEffect, useRef, useState } from 'react';
import {
  Animated,
  Easing,
  Image,
  ImageSourcePropType,
  LayoutChangeEvent,
  StyleSheet,
  View,
  ViewStyle,
} from 'react-native';
import { useTheme } from 'react-native-paper';

import WelcomeText from './WelcomeText';

import DefaultProfilePic from 'assets/images/profile-default.jpeg';

export type UserData = null | {
  name: string
  image?: ImageSourcePropType
}

type ProfileBarProps = {
  style?: ViewStyle
  user: UserData
}

export default function ProfileBar({ style, user }: ProfileBarProps) {
  const [viewWidth, setViewWidth] = useState(0);
  const fadeInAnim = useRef(new Animated.Value(0)).current;
  const theme = useTheme();
  const styles = useStyles(theme);
  const viewPadding = 5;
  const avatarSize = 50;
  const translateX = fadeInAnim.interpolate({
    inputRange: [0, 1],
    outputRange: [viewWidth - avatarSize - viewPadding * 2, 0],
  });
  const rotate = fadeInAnim.interpolate({
    inputRange: [0, 1],
    outputRange: ['0deg', '360deg'],
  });

  const onLayout = (event: LayoutChangeEvent) => {
    const { layout } = event.nativeEvent;
    setViewWidth(layout.width);
  };

  const runFadeInAnimation = () => {
    Animated.timing(fadeInAnim, {
      useNativeDriver: true,
      duration: 1000,
      toValue: 1,
      easing: Easing.bezier(0.16, 1, 0.3, 1),
    }).start();
  };

  useEffect(() => {
    runFadeInAnimation();
  }, []);

  const viewStyle = { padding: viewPadding, opacity: fadeInAnim };
  const picStyle = {
    transform: [{ translateX }, { rotate }],
    opacity: fadeInAnim,
  };

  return (
    <Animated.View
      style={[styles.wrap, style, viewStyle]}
      onLayout={onLayout}
    >
      <View style={styles.row}>
        <Animated.View style={[styles.avatar, picStyle]}>
          <Image
            source={user && user.image ? user.image : DefaultProfilePic}
            style={styles.image}
            resizeMode="cover"
          />
        </Animated.View>
        <Animated.View style={[styles.content, { opacity: fadeInAnim }]}>
          <WelcomeText user={user} theme={theme} />
        </Animated.View>
      </View>
    </Animated.View>
  );
}

const useStyles = ({ colors }: ReactNativePaper.Theme) => StyleSheet.create({
  wrap: {
    width: '100%',
    padding: 5,
    backgroundColor: colors.darkGrey,
    borderRadius: 64,
  },
  row: {
    flexDirection: 'row',
  },
  avatar: {
    width: 50,
    height: 50,
  },
  image: {
    width: '100%',
    height: '100%',
    borderRadius: 50,
  },
  content: {
    flex: 1,
    paddingHorizontal: 15,
    paddingRight: 20,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    textAlign: 'center',
  },
  accent: {
    color: colors.accent,
    fontWeight: 'bold',
  },
});
