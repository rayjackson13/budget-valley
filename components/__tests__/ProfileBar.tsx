import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import ProfileImage from 'assets/images/profile-default.jpeg';
import ProfileBar from '../ProfileBar';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

const mockUser = { image: ProfileImage, name: 'Kostya' };

it('renders correctly', () => {
  const tree = renderer.create(<ProfileBar user={mockUser} />).toJSON();
  expect(tree).not.toBe(null);
});

it('shows current profile', () => {
  const { queryByTestId, toJSON } = render(<ProfileBar user={mockUser} />);
  const userNameNode = queryByTestId('username');
  expect(userNameNode).not.toBe(null);
  expect(userNameNode?.props.children).toBe(mockUser.name);
  expect(toJSON()).toMatchSnapshot();
});
