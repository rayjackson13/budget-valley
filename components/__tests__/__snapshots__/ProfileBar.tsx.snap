// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows current profile 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": undefined,
      "borderRadius": 64,
      "opacity": 0,
      "padding": 5,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "height": 50,
          "opacity": 0,
          "transform": Array [
            Object {
              "translateX": -60,
            },
            Object {
              "rotate": "0deg",
            },
          ],
          "width": 50,
        }
      }
    >
      <Image
        resizeMode="cover"
        source={1}
        style={
          Object {
            "borderRadius": 50,
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 0,
          "paddingHorizontal": 15,
          "paddingRight": 20,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "500",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                Object {
                  "textAlign": "center",
                },
              ],
            ],
          ]
        }
      >
        Welcome back, 
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "#03dac4",
                "fontWeight": "bold",
              },
            ]
          }
          testID="username"
        >
          Kostya
        </Text>
        !
      </Text>
    </View>
  </View>
</View>
`;
